src/compiler/checker.ts

```diff
@@ -18258,6 +18251,9 @@ namespace ts {
         function checkExpressionCached(node: Expression, checkMode?: CheckMode): Type {
             const links = getNodeLinks(node);
             if (!links.resolvedType) {
+                if (checkMode) {

```

Comment:
@weswigham I'd like to understand the motivation for the change as well.

@sandersn Caching isn't cost free (i.e. it will often cause allocation of a `NodeLinks` object) so we don't want to *always* cache.

Reply:
I wanted to cache results in `createJsxAttributesTypeFromAttributesProperty` (as it's called at least twice for each overload resolved through JSX), but it _could_ be called in a contextual context (and often is), in which case caching indeterminate results would be bad. I'm not in for putting ad-hoc checks on cache safety at every use-site - I'd rather `checkExpressionCached` make it impossible to do the wrong thing.

Reply:
Oh, @ahejlsberg if we want to reduce the cost of allocation of our `NodeLinks` structure (do we? we totally could), we can replace the "A cache of structures which link nodes to other things" structure with "a collection of caches which maps nodes directly to things". So instead of allocating a `NodeLinks` on each node we cache something on, we initially allocate a `Map` for each of the current fields of `NodeLinks`, and use the nodeid as the key on those directly. No extra object allocations, just extra references. Probably also update usages to `getNodeLink(node, "propName")` and `setNodeLink(node, "propName")`, where `getNodeLink` is implemented like
```ts
function getNodeLink<K extends keyof NodeLinks>(node: Node, prop: K): NodeLinks[K] {
  switch (prop) {
    case "resolvedType":
      return resolvedTypes.get(getNodeId(node));
    // ... and so on
  }
}
```
with our modern features this is completely typesafe, too. 🐱